File: docs/scripts/DumpNcsexperSyscallTable.java
// Dump NCSEXPER's IPO interpreter syscall table.
//
// Output is written to a file (configurable below) so you can paste it
// back into the chat without the Console window truncating it.
//
// The filter is tightened to runs of 100..300 consecutive .text pointers —
// the typical size of an interpreter dispatch table (~160 entries by inpax
// reckoning). Tighter than the original "≥16" gate that caught every C++
// vtable and switch table in the binary.

//@category NCSExpert
//@author ncsx
//@menupath Tools.NCSExpert.Dump syscall table
//@toolbar globe.png

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.symbol.Symbol;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;

public class DumpNcsexperSyscallTable extends GhidraScript {

    // .text segment range (from NCSEXPER.EXE's PE headers — verified).
    private static final long TEXT_START = 0x00401000L;
    private static final long TEXT_END   = 0x005ad9ffL;

    // .rdata segment range.
    private static final long RDATA_START = 0x005ae000L;
    private static final long RDATA_END   = 0x006017ffL;

    // Candidate-run filter: the IPO interpreter syscall table is expected to
    // be ~160 entries; vtables are smaller, big tables (image-id arrays etc.)
    // are larger. Tune the range here if no candidates surface.
    private static final int MIN_RUN_LEN = 100;
    private static final int MAX_RUN_LEN = 300;

    // Output file — drop the script result here. Adjust if you want it
    // somewhere else; the path is platform-agnostic (Java handles slashes).
    private static final String OUTPUT_PATH =
        "/Users/mjaskols/Projects/my/ncsx/docs/scripts/syscall-table-dump.txt";

    // Anchor functions we already located, for sanity-tagging candidates.
    private static final long DO_INTERPRET_ADDR     = 0x0045d830L;
    private static final long SYSCALL_RESOLVER_ADDR = 0x004689e0L;

    @Override
    public void run() throws Exception {
        BufferedWriter w = new BufferedWriter(new FileWriter(OUTPUT_PATH));
        try {
            writeLine(w, "================================================================");
            writeLine(w, "NCSEXPER syscall table scan");
            writeLine(w, "================================================================");
            writeLine(w, "Scanning  .rdata 0x" + hex(RDATA_START) + " .. 0x" + hex(RDATA_END));
            writeLine(w, "Filtering .text  0x" + hex(TEXT_START) + " .. 0x" + hex(TEXT_END));
            writeLine(w, "Run-length filter: " + MIN_RUN_LEN + " .. " + MAX_RUN_LEN);
            writeLine(w, "Anchor: DoInterpret 0x" + hex(DO_INTERPRET_ADDR)
                       + ", syscall resolver 0x" + hex(SYSCALL_RESOLVER_ADDR));
            writeLine(w, "");

            // Walk .rdata 4 bytes at a time, tracking runs of consecutive code pointers.
            long addr = RDATA_START;
            int candidateCount = 0;

            while (addr < RDATA_END - 64) {
                long runStart = addr;
                int runLen = 0;
                while (addr + 4 <= RDATA_END && isTextAddr(readU32(addr))) {
                    addr += 4;
                    runLen++;
                }
                if (runLen >= MIN_RUN_LEN && runLen <= MAX_RUN_LEN) {
                    candidateCount++;
                    dumpCandidate(w, runStart, runLen, candidateCount);
                } else if (runLen == 0) {
                    addr += 4;
                }
                // (otherwise `addr` already past the run; loop continues)
            }

            writeLine(w, "");
            writeLine(w, "================================================================");
            writeLine(w, "Done. " + candidateCount + " candidate(s) in the "
                          + MIN_RUN_LEN + ".." + MAX_RUN_LEN + " range.");
            writeLine(w, "================================================================");

            println("Wrote " + candidateCount + " candidate(s) to:");
            println("  " + OUTPUT_PATH);
        } finally {
            w.close();
        }
    }

    private void dumpCandidate(BufferedWriter w, long start, int runLen, int idx)
        throws IOException
    {
        writeLine(w, "");
        writeLine(w, "----------------------------------------------------------------");
        writeLine(w, "CANDIDATE #" + idx + " @ 0x" + hex(start) + "  (" + runLen + " entries)");
        writeLine(w, "----------------------------------------------------------------");
        for (int i = 0; i < runLen; i++) {
            long ptrAddr = start + ((long) i) * 4L;
            long fnAddr = readU32(ptrAddr);
            String fnName = nameOf(fnAddr);
            writeLine(w, String.format("  [0x%02x] = 0x%08x  (%s)", i, fnAddr, fnName));
        }
    }

    private void writeLine(BufferedWriter w, String s) throws IOException {
        w.write(s);
        w.newLine();
    }

    private long readU32(long offset) {
        Address a = toAddr(offset);
        if (a == null) return 0L;
        try {
            return ((long) getInt(a)) & 0xFFFFFFFFL;
        } catch (Exception e) {
            return 0L;
        }
    }

    private boolean isTextAddr(long value) {
        return value >= TEXT_START && value <= TEXT_END;
    }

    private String nameOf(long fnAddr) {
        Address a = toAddr(fnAddr);
        if (a == null) return "<invalid addr>";
        Function fn = getFunctionAt(a);
        if (fn != null) return fn.getName();
        Symbol sym = getSymbolAt(a);
        if (sym != null) return sym.getName();
        return String.format("FUN_%08x", fnAddr);
    }

    private static String hex(long v) {
        return String.format("%08x", v & 0xFFFFFFFFL);
    }
}

File: docs/scripts/dump-ncsexper-syscall-table.py
# Dump NCSEXPER's IPO interpreter syscall table.
#
# Paste into Ghidra's Script Manager (File → Configure → Script Manager → New)
# with NCSEXPER.EXE loaded. Or run via `analyzeHeadless` if scripting from CLI.
#
# Strategy:
#
# 1. Locate the syscall resolver (FUN_004689e0). It reads the table base via
#    `*(this + 0x50)` and indexes by ID.
# 2. Find any caller of that resolver — that gives us a code location where
#    the `this` pointer is set. The `this` is usually a CInterpreter
#    singleton stored at a known global.
# 3. Walk all reads of `this + 0x50` to find what static address ends up there
#    (i.e. find a constructor that does `mov [ecx+0x50], <static_table_addr>`).
# 4. From the static table address, read N function pointers (each 4 bytes,
#    little-endian) until the values stop looking like .text addresses.
# 5. For each entry, look up the function it points at and dump
#    `[id] = 0xADDR  (function_name)`.
#
# Output format is paste-ready for incorporation into
# `packages/inpax-cabi-provider/src/ncsexper-syscalls.ts` — one line per slot.

from ghidra.program.model.address import AddressSet
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

# ─── Known anchor points (from prior ghidra trace) ─────────────────────────
SYSCALL_RESOLVER_ADDR = 0x004689e0   # FUN_004689e0 — resolves (flag, id) → handler*
DO_INTERPRET_ADDR     = 0x0045d830   # FUN_0045d830 — CInterpreter::DoInterpret
TEXT_START            = 0x00401000   # .text segment start
TEXT_END              = 0x005ad9ff   # .text segment end

# ─── Step 1: Find what static address gets stored at `this+0x50` ───────────
#
# We scan the .text segment for an instruction sequence that writes a 4-byte
# value into [reg + 0x50]. Common constructor pattern:
#     mov  ecx, this
#     mov  [ecx + 0x50], <addr_of_table>
#     mov  [ecx + 0x54], <table_size>
#
# Because pinning the exact constructor is brittle, we search by data: any
# 4-byte value in .rdata that points into .text AND is followed by more such
# values, is a function-pointer table candidate. The table that contains
# handlers for the IDs we know NCSEXPER calls (0x0B setjobstatus, 0x0C exit,
# 0x0D apiJob-bridge, 0x2E PEMSGZ_Kopfzeile, 0x33 PEMProtokollAusgabe, …) is
# our syscall table.

def is_text_addr(value):
    return TEXT_START <= value <= TEXT_END

def read_u32(addr):
    """Read little-endian uint32 at a given numeric address."""
    a = toAddr(addr)
    if a is None:
        return None
    try:
        return getInt(a) & 0xFFFFFFFF
    except Exception:
        return None

def get_func_name(addr):
    """Pretty-print the function (or label) at a given address."""
    a = toAddr(addr)
    if a is None:
        return "<invalid addr>"
    fn = getFunctionAt(a)
    if fn is not None:
        return fn.getName()
    sym = getSymbolAt(a)
    if sym is not None:
        return sym.getName()
    return "FUN_{:08x}".format(addr)

# ─── Step 2: scan .rdata for function-pointer tables ───────────────────────
#
# Stride 4 bytes. A run of >= 16 consecutive valid-text addresses is a
# candidate table. For each candidate, print the first 16 entries with their
# function names — you eyeball which one has the right shape (e.g. slot 0x0B
# should look like a job-status setter; slot 0x2E should look like a PEM
# header writer).

RDATA_START = 0x005ae000
RDATA_END   = 0x006017ff

print("Scanning .rdata for syscall-table candidates...")
print("(looking for runs of >= 16 consecutive function pointers)")
print()

addr = RDATA_START
candidates = []
while addr < RDATA_END - 64:
    run_start = addr
    run_len = 0
    while addr < RDATA_END and is_text_addr(read_u32(addr) or 0):
        addr += 4
        run_len += 1
    if run_len >= 16:
        candidates.append((run_start, run_len))
    else:
        addr += 4

print("Found {} candidate runs:".format(len(candidates)))
for (start, run_len) in candidates:
    print("  0x{:08x}: {} entries".format(start, run_len))
print()

# ─── Step 3: dump each candidate ───────────────────────────────────────────
#
# Print the first up-to-100 entries of each candidate with function names.
# Manually identify which run is the syscall table — it should have:
#
#   [0x00] = setmenutitle-like     (one-arg string setter)
#   [0x0B] = setjobstatus-like     (one-arg string setter)
#   [0x0C] = exit                  (zero-arg)
#   [0x0D] = apiJob bridge         (four-arg, references api32.dll)
#   [0x2E] = PEM_header-like
#   ... etc.

for (start, run_len) in candidates:
    print("─" * 72)
    print("CANDIDATE @ 0x{:08x}  ({} entries)".format(start, run_len))
    print("─" * 72)
    n = min(run_len, 100)
    for i in range(n):
        ptr_addr = start + i * 4
        fn_addr = read_u32(ptr_addr)
        name = get_func_name(fn_addr)
        print("  [0x{:02x}] = 0x{:08x}  ({})".format(i, fn_addr, name))
    print()

print("done.")
print()
print("→ Paste the matching candidate into")
print("  packages/inpax-cabi-provider/src/ncsexper-syscalls.ts")
print("  upgrading each entry's `verified` to either")
print("  'verified-same' or 'divergent'.")
